libraries/freertos_plus/standard/https/src/iot_https_client.c

```diff
@@ -0,0 +1,1829 @@
+    if(_httpsResponse->bufferProcessingState == PROCESSING_STATE_SEARCHING_HEADER_BUFFER)
+    {
+        if(_httpsResponse->foundHeaderField)
+        {
+            _httpsResponse->pReadHeaderValue = ( uint8_t* )( pLoc );
+            _httpsResponse->readHeaderValueLength = length;
+            /* We found a header field so we don't want to keep parsing.*/
+            return 1;
+        }
+    }
+    return 0;
+}
+
+/*-----------------------------------------------------------*/
+
+static int _httpParserOnHeadersCompleteCallback(http_parser * pHttpParser)
+{
+    IotLogDebug("Parser: End of the headers reached in the pHttpParser.");
+    _httpsResponse_t * _httpsResponse = (_httpsResponse_t *)(pHttpParser->data);
+    _httpsResponse->parserState = PARSER_STATE_HEADERS_COMPLETE;
+
+    /* If the IotHttpsClient_ReadHeader() was called, we return after finishing looking through all of the headers. 
+       Returning a non-zero value exists the http parsing. */
+    if(_httpsResponse->bufferProcessingState == PROCESSING_STATE_SEARCHING_HEADER_BUFFER)
+    {
+        return 1;
+    }
+
+    /* When in this callback the pheaderCur pointer is at the first "\r" in the last header line. HTTP/1.1
+       headers end with another "\r\n" at the end of the last line. This means we must increment
+       the headerCur pointer to the length of "\r\n\r\n". */
+    if(_httpsResponse->bufferProcessingState == PROCESSING_STATE_FILLING_HEADER_BUFFER)
+    {
+        _httpsResponse->pHeadersCur += strlen("\r\n\r\n");
+    }
+
+    /* content_length will be zero if no Content-Length header found by the parser. */
+    _httpsResponse->contentLength = pHttpParser->content_length;
+
+    /* For a HEAD method, there is no body expected in the response, so we return 1 to skip body parsing. 
+       Also if it was configured in a synchronous response to ignore the HTTPS response body then also stop the body 
+       parsing. */
+    if((_httpsResponse->method == IOT_HTTPS_METHOD_HEAD) || (_httpsResponse->pBody == NULL))
+    {
+        return 1;
+    }
+    return 0;
+}
+
+/*-----------------------------------------------------------*/
+
+static int _httpParserOnBodyCallback(http_parser * pHttpParser, const char * pLoc, size_t length)
+{
+    IotLogDebug("Parser: Reached the HTTPS message body. It is of length: %d", length);
+
+    _httpsResponse_t * _httpsResponse = (_httpsResponse_t *)(pHttpParser->data);
+    _httpsResponse->parserState = PARSER_STATE_IN_BODY;
+
+    /* If we are in the header buffer and we come across HTTPS body, then we want to copy it to the body
+       buffer. */
+    /* Only copy the data if the current location is not the bodyCur. */
+    if((_httpsResponse->pBodyCur != NULL) && (_httpsResponse->pBodyEnd - _httpsResponse->pBodyCur > 0))
+    {
+        if(_httpsResponse->pBodyCur != pLoc)
+        {
+            memcpy(_httpsResponse->pBodyCur, pLoc, length);
```

Comment:
WHats the guarantee that ```length``` amount of bytes can be stored in ```_httpsResponse->pBodyEnd - _httpsResponse->pBodyCur``` ?

Reply:
The guaranteed is because parser is parsing only the data within the bounds of the buffer.
See:
```
static IotHttpsReturnCode_t _parseHttpsMessage(http_parser* pHttpParser, char* pBuf, size_t len)
{
    size_t parsedBytes = http_parser_execute(pHttpParser, &_httpParserSettings, pBuf, len );
```